gen_vimdoc.py: add more structure to mpack result

smolck opened the pull request:
See TODO in https://github.com/neovim/neovim/commit/46bde66147ead1bb234932a48e2c56fc4b698ec0.

Definitely isn't perfect, so any feedback is appreciated.

EDIT: There's likely to be some unnecessary code due to my re-purposing of the functions already in gen_vimdoc.py, so please let me know if you see any such code or functions.

smolck committed:
[scripts/gen_vimdoc.py] Generate better-formatted mpack

Changes the generated msgpack result values in the runtime/doc/*.mpack
files to be formatted like this (instead of being formatted like Vim help text):
    [
        'nvim_win_get_var': {
                'signature': 'nvim_win_get_var({window}, {name}, {err})',
                'parameters': [('Window', 'window'), ('String', 'name'), ('Error *', 'err')],
                'parameters_doc': [{'window': 'Window handle, or 0 for current window', 'name': 'Variable name'}],
                'doc': ['Gets a window-scoped (w:) variable'],
                'return': ['Variable value'],
                'seealso': []
            }
        },
        ...
    ]

smolck commented on scripts/gen_vimdoc.py:
```diff
@@ -365,6 +395,86 @@ def render_node(n, text, prefix='', indent='', width=62):
+        'params': Information from <parameterlist>'s
+        'return': TODO
+        'seealso': TODO
+        'xrefs': TODO
```
Not so sure on what to put for the docstrings here. Also, is `xrefs` necessary?

justinmk replied:
not sure. xrefs might be useful for docstrings that "link" to each other, but currently it's just there "for fun".

smolck commented on scripts/gen_vimdoc.py:
```diff
@@ -627,6 +868,7 @@ def gen_docs(config):
         if p.returncode:
             sys.exit(p.returncode)
 
+        mpacks = {}
```
Maybe need a different name for this?

smolck commented on scripts/gen_vimdoc.py:
```diff
@@ -443,13 +563,134 @@ def parse_parblock(parent, prefix='', width=62, indent=''):
 # }}}
 
 
+def parse_source_xml_for_mpack(filename, mode):
```
Different name?

smolck commented on scripts/gen_vimdoc.py:
```diff
@@ -262,6 +262,36 @@ def has_nonexcluded_params(nodes):
+        desc_node = get_child(node, 'parameterdescription')
+        if desc_node:
+            desc = parse_parblock(desc_node, width=width,
+                                  indent=(' ' * len(name)))
```
Should probably remove this `indent=` part?

justinmk committed:
gen_vimdoc.py: DRY

justinmk commented:
> There's likely to be some unnecessary code due to my re-purposing of the functions already in gen_vimdoc.py, so please let me know if you see any such code or functions

I pushed a commit which eliminates the extra `parse_source_xml()` variant. Somewhat hacky in that I added a new paramter `fmt_vimdoc`, but this is preferable to having many big-yet-similar functions.

Maybe the same treatment could be applied to the other cases?

justinmk commented on scripts/gen_vimdoc.py:
```diff
@@ -365,6 +396,86 @@ def render_node(n, text, prefix='', indent='', width=62):
+    """Parses XML <para> as a map
+
+    The map returned may or may not contain
+    all of the following keys:
```
Makes things a lot nicer if callers can always assume keys exist, no?

justinmk commented:
I regressed this a bit. Will fix later.

smolck commented on scripts/gen_vimdoc.py:
```diff
@@ -563,21 +564,23 @@ def parse_parblock(parent, prefix='', width=62, indent=''):
 
-    Caller decides what to do with the deprecated documentation.
+    The `fmt_vimhelp` parameter controls some special cases for use by
+    fmt_doxygen_xml_as_vimhelp(). (TODO: ugly :)
```
What do you think of building the Vim help text in `fmt_doxygen_xml_as_vimhelp()` directly from the function maps returned by `extract_from_xml`? That way there’s no need for the fmt_vimhelp and desc_node stuff. Only thing is that I’m unsure how much more work that’d be.

justinmk replied:
That would be ideal.

justinmk committed:
gen_vimdoc.py: fix mpack generator

- make parameters_doc a dict intead of a list
  BEFORE:
    "parameters_doc": [
        {
            "buffer": "Buffer handle, or 0 for current buffer"
        }
    ],
  AFTER:
    "parameters_doc": {
        "buffer": "Buffer handle, or 0 for current buffer"
    },
- make "return", "seealso", lists instead of strings

justinmk committed:
gen_vimdoc.py: mpack: exclude deprecated functions

The `mpack` variable was a tuple, which manifests as an array in the
generated msgpack structure.
- Removes noise from the mpack data (deprecated functions are
  deprecated).
- Eliminates 1 level of nesting.

BEFORE:
    [
      {
        "buffer.c": [
          {
            "nvim__buf_stats": { ... },
            ...
          },
          {
            "buffer_del_line": { ... },
            ...
          },
        ],
        ...
      }
    ]

AFTER:
    [
      {
        "buffer.c": {
          "nvim__buf_stats": { ... },
          ...
        },
        ...
    ]

justinmk committed:
gen_vimdoc.py: fix "seealso", "xrefs"

- Also fix xrefs ("Deprecated" section)
- Fix "Deprecated" rendering by a weird hack (see comment).
- Eliminate unnecessary use of render_para()

justinmk committed:
gen_vimdoc.py: mpack: collect functions in 1 dict

All Nvim API, core Vimscript, and core Lua functions are globally
unique, so there is no need for per-module nested dicts.

BEFORE (generated mpack structure):
    [
      {
        "buffer.c": {
          "nvim__buf_stats": { ... },
          ...
        },
        "window.c": {
          "nvim_win_close": { ... },
          ...
        },
        ...
      }
    ]

AFTER (generated mpack structure):
    [
      {
        "nvim__buf_stats": {
          ...
        },
        "nvim_buf_attach": {
          ...
        },
        "nvim_tabpage_set_var": {
          ...
        },
        "nvim_ui_attach": {
          ...
        },
        "nvim_win_close": {
          ...
        }
      }
    ]

justinmk committed:
gen_vimdoc.py: rename for clarity

- render_para => fmt_node_as_vimhelp
- Inline parse_parblock() in fmt_node_as_vimhelp()

justinmk commented:
@smolck  latest version produces pretty good `*.mpack` results. There are some formatting regressions for the Vim `:help` results, but I'm close to fixing that. Meanwhile I'll merge this. 

Remaining question for the "unformatted" (`*.mpack`) forms is how to represent lists and line breaks. Currently line breaks (and indent) are included in the `parameters_doc` map. It's easy enough for API clients to just strip linebreak-preceded whitespace, but that will break `<pre>` formatted blocks (and nested lists).

Here's an example from `api.mpack`:

```
"nvim_replace_termcodes": {
    "annotations": [],
    "doc": [
        "Replaces terminal codes and |keycodes| (<CR>, <Esc>, ...) in a\nstring with the internal representation."
    ],
    "parameters": [
        [
            "String",
            "str"
        ],
        [                      
            "Boolean",                                                   
            "from_part"                                   
        ],                                                           
        [
            "Boolean",                          
            "do_lt"           
        ],        
        [                                                             
            "Boolean",
            "special"         
        ]                
    ],           
    "parameters_doc": {                                                                                           
        "do_lt": "Also translate <lt>. Ignored if `special` is false.",                 
        "from_part": "Legacy Vim parameter. Usually true.",                 
        "special": "Replace |keycodes|, e.g. <CR> becomes a \"\\n\" char.",                                                  
        "str": "String to be converted."                                                                                                      
    },           
    "return": [],      
    "seealso": [
        "    replace_termcodes",                                                        
        "    cpoptions"
    ],                                                               
    "signature": "nvim_replace_termcodes({str}, {from_part}, {do_lt}, {special})"
}, 
```

smolck commented:
LGTM! Thanks for all the help with this!

justinmk commented:
@smolck Note that the mpack structure changed. Here's a 2-liner to preview it in human-readable form:

```
:new | put=json_encode(msgpackparse(readfile('runtime/doc/api.mpack')))
:%!python -m json.tool
```

blueyed commented:
btw: running `scripts/gen_vimdoc.py` changes `runtime/doc/api.txt` and `runtime/doc/lua.txt`.  Is that intentional?

justinmk commented:
@blueyed that's what I referred to here:

> There are some formatting regressions for the Vim `:help` results, but I'm close to fixing that.
